### ISSUE TITLE
CNNT assumes fixed patch count causing positional embedding mismatch for different input sizes

### ISSUE THREAD

**Issue by SarthakJagota** [ML4SCI/DeepLense#149]
The current CNNT implementation uses a hard-coded patch count to initialize positional embeddings. This assumes a fixed input resolution and can lead to shape mismatches when training or evaluating on datasets with different image sizes.

Since the number of patches depends on the CNN feature map, positional embeddings should be created dynamically based on the actual patch sequence length.

Problems

1. Model breaks for variable input resolutions

2. Limits reuse across DeepLense tasks

3. Can cause runtime shape mismatch errors

Proposed solution

1. Remove the hard-coded patch count

2. Create positional embeddings dynamically using the CNN output           patch length

3. Properly register the parameter so it is tracked by the optimizer

This would make CNNT resolution-agnostic and improve flexibility across datasets.

**Comment by Apprentice2907** [comment]
Thanks for the fix! The dynamic positional embedding idea makes sense for handling variable input sizes.

One question: since `nn.Parameter` is being created inside `forward()`, could this cause issues with optimizer state or checkpoint loading? Typically PyTorch parameters are initialized in `__init__`.

Would it make sense to instead initialize a base positional embedding and interpolate or resize it dynamically during the forward pass?

Happy to help test or explore alternatives if needed.

**Comment by Akarshak51** [comment]
The concern about creating `nn.Parameter` in `forward()` is valid and represents a fundamental PyTorch anti-pattern. Here's a robust solution:

**Root Cause:**
The CNNT architecture hard-codes positional embedding dimensions at initialization, making it incompatible with variable CNN feature map sizes that differ across datasets or input resolutions.

**Recommended Implementation:**

```python
class CNNT(nn.Module):
    def __init__(self, max_patches=256):
        super().__init__()
        # Initialize with maximum expected patch count
        self.max_patches = max_patches
        self.pos_embedding = nn.Parameter(torch.randn(1, max_patches, embed_dim))
        
    def forward(self, x):
        B, N, D = x.shape  # N = actual patch count from CNN
        
        if N > self.max_patches:
            # Extend via interpolation if needed
            pos_emb = F.interpolate(
                self.pos_embedding.permute(0, 2, 1),
                size=N,
                mode='linear',
                align_corners=False
            ).permute(0, 2, 1)
        else:
            # Slice to match actual patch count
            pos_emb = self.pos_embedding[:, :N, :]
            
        return x + pos_emb.expand(B, -1, -1)
```

**Why This Works:**
- Parameters initialized in `__init__` are properly tracked by optimizer and state_dict
- Interpolation handles arbitrary input sizes without creating new parameters
- Slicing for smaller inputs is memory-efficient
- Backward pass correctly accumulates gradients only to the base parameter

**Alternative - Sinusoidal Embeddings:**
For true resolution-agnostic behavior without interpolation artifacts:

```python
def get_sinusoidal_embeddings(seq_len, dim):
    position = torch.arange(seq_len).unsqueeze(1)
    div_term = torch.exp(torch.arange(0, dim, 2) * -(math.log(10000.0) / dim))
    pos_emb = torch.zeros(seq_len, dim)
    pos_emb[:, 0::2] = torch.sin(position * div_term)
    pos_emb[:, 1::2] = torch.cos(position * div_term)
    return pos_emb.unsqueeze(0)
```

This avoids learnable parameters entirely while maintaining translation equivariance properties.

Would be happy to help test either approach or validate against the DeepLense benchmark datasets.

**Comment by vignathi123-vi** [comment]
Hi! I've submitted PR to fix this issue.
The root cause was num_patches = 16 * 16 being hardcoded in CNNT's 
__init__, which assumes a fixed input resolution and causes shape 
mismatches when the CNN feature map produces a different patch count.

The fix:
 Replaces hardcoded num_patches with max_patches = 256
 For inputs smaller than max_patches → positional embedding is sliced 
   dynamically via pos_embedding[:, :(n + 1)]
 For inputs larger than max_patches → interpolation is used to extend 
   the positional embedding to the required size
This makes CNNT truly resolution-agnostic and compatible across all 
DeepLense datasets and input sizes. 
Happy to make adjustments if needed!
